File: design_patterns/002_observer/solutions/python/weatherStation/src/DisplayForecast.py
from IDisplay import IDisplay
from IObserver import IObserver
from IPublisher import IPublisher
from WeatherData import WeatherInfo


class DisplayForecast(IObserver, IDisplay):
    _weather_data:IPublisher = None
    _current_pressure:float = 29.92
    _last_pressure:float = None

    def __init__(self, weather_data: IPublisher) -> None:
        self._weather_data = weather_data
        self._weather_data.subscribe(self)


    def update(self, updated_data: dict) -> None:
        if current_pressure := updated_data.get(WeatherInfo.PRESSURE):
            self._last_pressure = self._current_pressure
            self._current_pressure = current_pressure
            self.display()


    def display(self) -> None:
        if self._current_pressure > self._last_pressure:
            forecast = 'Improving weather on the way!'
        elif self._current_pressure == self._last_pressure:
            forecast = 'More of the same'
        else:
            forecast = 'Watch out for cooler, rainy weather'
        print(f'Forecast: {forecast}')


if __name__ == '__main__':
    raise Exception(f'This module should not be executed directly. Only for imports')

File: design_patterns/002_observer/solutions/typescript/weatherStation/src/DisplayForecast.ts
import { IDisplay } from './IDisplay';
import { IObserver } from './IObserver';
import { IPublisher } from './IPublisher';
import { WeatherInfo } from './WeatherData';

export class DisplayForecast implements IObserver, IDisplay {
  private _weather_data: IPublisher | null = null;
  private _current_pressure: number = 29.92;
  private _last_pressure: number | null = null;

  constructor(weather_data: IPublisher) {
    this._weather_data = weather_data;
    this._weather_data.subscribe(this);
  }

  update(updated_data: { [key: string]: any }): void {
    const current_pressure = updated_data[WeatherInfo.PRESSURE];
    if (current_pressure !== undefined) {
      this._last_pressure = this._current_pressure;
      this._current_pressure = current_pressure;
      this.display();
    }
  }

  display(): void {
    let forecast: string;
    if (this._last_pressure === null) {
      forecast = 'No data for comparison';
    } else if (this._current_pressure > this._last_pressure) {
      forecast = 'Improving weather on the way!';
    } else if (this._current_pressure === this._last_pressure) {
      forecast = 'More of the same';
    } else {
      forecast = 'Watch out for cooler, rainy weather';
    }
    console.log(`Forecast: ${forecast}`);
  }
}
